ExtractUserDataActivity/handler.ts

```diff
@@ -0,0 +1,526 @@
+  ActivityResultFailure
+]);
+export type ActivityResult = t.TypeOf<typeof ActivityResult>;
+
+const logPrefix = `ExtractUserDataActivity`;
+
+/**
+ * Converts a Promise<Either> into a TaskEither
+ * This is needed because our models return unconvenient type. Both left and rejection cases are handled as a TaskEither left
+ * @param lazyPromise a lazy promise to convert
+ * @param queryName an optional name for the query, for logging purpose
+ *
+ * @returns either the query result or a query failure
+ */
+const fromQueryEither = <R>(
+  lazyPromise: () => Promise<Either<QueryError | Error, R>>,
+  queryName: string = ""
+): TaskEither<ActivityResultQueryFailure, R> =>
+  tryCatch(lazyPromise, (err: Error) =>
+    ActivityResultQueryFailure.encode({
+      kind: "QUERY_FAILURE",
+      query: queryName,
+      reason: err.message
+    })
+  ).chain((queryErrorOrRecord: Either<QueryError | Error, R>) =>
+    fromEither(
+      queryErrorOrRecord.mapLeft(queryError =>
+        ActivityResultQueryFailure.encode({
+          kind: "QUERY_FAILURE",
+          query: queryName,
+          reason: JSON.stringify(queryError)
+        })
+      )
+    )
+  );
+
+/**
+ * To be used for exhaustive checks
+ * @param _
+ */
+function assertNever(_: never): void {
+  throw new Error("should not have executed this");
+}
+
+/**
+ * Logs depending on failure type
+ * @param context the Azure functions context
+ * @param failure the failure to log
+ */
+const logFailure = (context: Context) => (
+  failure: ActivityResultFailure
+): void => {
+  switch (failure.kind) {
+    case "INVALID_INPUT_FAILURE":
+      context.log.error(
+        `${logPrefix}|Error decoding input|ERROR=${failure.reason}`
+      );
+      break;
+    case "QUERY_FAILURE":
+      context.log.error(
+        `${logPrefix}|Error ${failure.query} query error |ERROR=${failure.reason}`
+      );
+      break;
+    case "USER_NOT_FOUND_FAILURE":
+      context.log.error(`${logPrefix}|Error user not found |ERROR=`);
+      break;
+    default:
+      assertNever(failure);
+  }
+};
+
+/**
+ * Skims a db document from db-related fields
+ * @param doc the document as retrieved from db
+ *
+ * @returns the same document without db metadata
+ */
+const fromRetrievedDbDocument = <T>(doc: T & RetrievedDocumentT): T => {
+  const RETRIEVED_DOCUMENT_KEYS: ReadonlyArray<keyof RetrievedDocumentT> = [
+    "_self",
+    "_ts",
+    "_attachments",
+    "_etag",
+    "_rid"
+  ];
+  return Object.entries(doc).reduce(
+    (p: T, [key, value]) =>
+      RETRIEVED_DOCUMENT_KEYS.includes(key) ? p : { ...p, [key]: value },
+    {} as T
+  );
+};
+
+/**
+ * Safely compose a message with a messa with its content
+ * @param messageFromDb a message as it's stored in db
+ * @param content a content for the message
+ *
+ * @returns a new message with content included
+ */
+const appendContentToMessage = (
+  messageFromDb: RetrievedMessageWithoutContent,
+  content: MessageContent
+): MessageWithContent => ({
```

Comment:
why do we need to merge the message with its content ?
it's better to just save them in a separate blob

Reply:
good, even easier

Reply:
the structure of the zip should mimic the internal representation (that won't happen, but suppose you have to restore all data fom the content of the zip :)

Reply:
This includes also document-related data? I mean: should we export also `RetrievedDocument`-related fields?

Reply:
no since they're not needed to rebuild the whole thing